sos/keyring.py

```diff
@@ -0,0 +1,453 @@
+
+    def get_name_from_id(self, keyid):
+        ret, out, err = self._system(["keyctl",
+                                      "rdescribe", str(keyid)], check=False)
+
+        if ret != 0:
+            raise KeyNotFoundError(keyid=keyid)
+
+        name = b";".join(out.split(b";")[4:])
+
+        return name.rstrip(b"\n")
+
+    def get_data_from_id(self, keyid, mode="raw"):
+        if mode.lower() == "raw":
+            kmode = "pipe"
+        elif mode.lower() == "hex":
+            kmode = "read"
+        else:
+            raise AttributeError("mode must be one of ['raw', 'hex'].")
+
+        ret, out, err = self._system(["keyctl",
+                                      kmode, str(keyid)], check=False)
+
+        if ret == 1:
+            raise KeyNotFoundError(keyid=keyid)
+
+        if mode == "raw":
+            return out
+        else:
+            h = b"".join(out.splitlines()[1:])
+            return h.replace(b" ", b"")
+
+    def get_custom_keyring_id(self):
+        ret, out, err = self._system(
+            ["keyctl", "describe", str(self.keyring)], check=False
+        )
+
+        if ret != 0:
+            raise KeyNotFoundError(keyname=self.keyring)
+
+        return out.decode().split(":")[0]
+
+    def add_key(self, name, data):
+        try:
+            keyid = self.get_id_from_name(name)
+            raise KeyAlreadyExistError(keyid=keyid, keyname=name)
+        except KeyNotFoundError:
+            pass
+
+        out = self._system(["keyctl", "padd",
+                            self.keytype, name, self.keyring], data)
+        keyid = int(out)
+
+        return keyid
+
+    def update_key(self, keyid, data):
+        ret, out, err = self._system(
+            ["keyctl", "pupdate", str(keyid)], data, check=False
+        )
+
+        if ret == 1:
+            raise KeyNotFoundError(keyid=keyid)
+        elif ret != 0:
+            raise KeyctlOperationError(keyid=keyid,
+                                       errmsg="({}){}".format(ret, err))
+
+    def remove_key(self, keyid: int):
+        ret, out, err = self._system(["keyctl",
+                                      "revoke", str(keyid)], check=False)
+        if ret == 1:
+            raise KeyNotFoundError(keyid=keyid)
+        elif ret != 0:
+            raise KeyctlOperationError(keyid=keyid,
+                                       errmsg="({}){}".format(ret, err))
+
+        self._system(["keyctl", "unlink", str(keyid), self.keyring])
+
+    def set_timeout(self, keyid, time):
+        ret, out, err = self._system(
+            ["keyctl", "timeout", str(keyid), str(time)], check=False
+        )
+
+        if ret == 1:
+            raise KeyNotFoundError(keyid=keyid)
+        elif ret != 0:
+            raise KeyctlOperationError(keyid=keyid,
+                                       errmsg="({}){}".format(ret, err))
+
+    def create_keyring(self, new_keyring_name, keyring=DEFAULT_KEYRING):
+        new_keyring_name = new_keyring_name.split(":")[-1]
```

Comment:
See above - this is a situation that should not happen, because we cannot create new rings from the user point of view. So no need for this code at all, and will be out in the next versionl